Use model generated by V2 in extended_check

rahxephon89 opened the pull request:
## Description
<!-- Please include a summary of the change, including which issue it fixes or what feature it adds. Include relevant motivation, context and documentation as appropriate. List dependencies that are required for this change, if any. -->

Currently, V2 compiler uses the model generated by V1 for the extended_check. This PR adds the logic to attach compiled module to the move model generated by V2 so that it can be used for the extended check.

close #12914


## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [x] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## How Has This Been Tested?
<!--
- Please ensure that the functionality introduced by this change is well tested and verified to work as expected.
- Ensure tests cover both happy and unhappy paths.
- List and link relevant tests.
-->

Existing tests still pass. 

The following flamegraphs are generated by running "aptos move compile --compiler-version=v2" from the folder `aptos-move/framework/aptos-framework`

flamegraph before change:
![flamegraph](https://github.com/aptos-labs/aptos-core/assets/10296575/bc4086c2-ff05-4476-a984-12ee7d28c693)


flamegraph after change:

![flamegraph](https://github.com/aptos-labs/aptos-core/assets/10296575/00c49683-3223-4ebb-9f8e-6eb8ef229110)

Execution time before change: 1.2s.
Execution time after change: 0.7s.


## Key Areas to Review
<!--
- Identify any critical parts of the code that require special attention or understanding. Explain why these parts are crucial to the functionality or architecture of the project.
- Point out any areas where complex logic has been implemented. Provide a brief explanation of the logic and your approach to make it easier for reviewers to follow.
- Highlight any areas where you are particularly concerned or unsure about the code's impact on the change. This can include potential performance or security issues, or compatibility with existing features.
-->

The logic to attach the compiled module to move model when generating the bytecode. 
## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

<!-- Thank you for your contribution! -->

rahxephon89 commented on third_party/move/tools/move-package/src/compilation/compiled_package.rs:
```diff
@@ -747,13 +747,28 @@ impl CompiledPackage {
-                flags,
-                &known_attributes,
-            )?;
+            let model = if resolution_graph.build_options.generate_docs
```
Still use V1 generated model for docgen and abigen

brmataptos replied:
Why is this necessary?

rahxephon89 replied:
When V2 model is used for docgen, links generated for modules from other packages are not correct because they are treated as targets. See the code [here](https://github.com/aptos-labs/aptos-core/blob/a72ef8a716ecf3ab207c8377cb94c9c5aedaf5b4/third_party/move/move-prover/move-docgen/src/docgen.rs#L470). abigen is also excluded in case I miss anything.

vineethk replied:
Please add appropriate comments here explaining this conditional check.

brmataptos commented:
> "aptos move compile --compiler-version=v2"
Where are you running this?  (For which package?)

rahxephon89 commented:
> > "aptos move compile --compiler-version=v2"
> > Where are you running this?  (For which package?)

From "aptos-move/framework/aptos-framework"

vineethk commented on third_party/move/tools/move-package/src/compilation/compiled_package.rs:
```diff
@@ -747,13 +747,28 @@ impl CompiledPackage {
+                // Use V2 generated model if not for docgen or abigen
+                env
+            } else {
+                run_model_builder_with_options_and_compilation_flags(
```
Looks like this code is identical to the `if` body. Can we please refactor the code to remove code dup (eg., check if `optional_global_env` is `None` in the if conditional, or similar).

rahxephon89 replied:
done

vineethk reviewed (approved):
Nice fix! Approving assuming comments will be addressed.

wrwg commented on third_party/move/move-compiler-v2/src/file_format_generator/mod.rs:
```diff
@@ -106,6 +121,12 @@ pub fn generate_file_format(
         }
     }
+    for (id, (m, map)) in module_data {
+        env.attach_compiled_module(id, m, map)
```
This really should not happen here: in the normal code path, we do not need to attach bytecode. It also duplicates logic which is already here: https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/move-compiler-v2/src/lib.rs#L142

rahxephon89 replied:
Added a experiment flag to control when to attach compiled modules.

wrwg reviewed (commented):
I'm not 100% sure what the purpose of the PR is. We already use the v2 model in extended checks. Its created here:

https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L1113-L1124

... which is called here:

https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/build_plan.rs#L139

... which in turn is called from aptos-move `BuiltPackage::build` which calls extended checks with the model as created by the v2 compiler:

https://github.com/aptos-labs/aptos-core/blob/219d461f7d4e0366bb802ac6282e6f0d8b7266ce/aptos-move/framework/src/built_package.rs#L207-L212

So this was already running in many e2e tests, and there is no building of the model twice since I fixed this a while ago and implemented passing the model back from.

Indeed this call chain does not attach bytecode, but is it needed? At least none of our v2 e2e tests failed which should take a similar code path.

If we really need to attach the bytecode to the model for running extended checks, we should add an option for doing so, and leverage the already existing code for this which I pointed to in the comment below.

rahxephon89 commented:
> I'm not 100% sure what the purpose of the PR is. We already use the v2 model in extended checks. Its created here:
> 
> https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L1113-L1124
> 
> ... which is called here:
> 
> https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/build_plan.rs#L139
> 
> ... which in turn is called from aptos-move `BuiltPackage::build` which calls extended checks with the model as created by the v2 compiler:
> 
> https://github.com/aptos-labs/aptos-core/blob/219d461f7d4e0366bb802ac6282e6f0d8b7266ce/aptos-move/framework/src/built_package.rs#L207-L212
> 
> So this was already running in many e2e tests, and there is no building of the model twice since I fixed this a while ago and implemented passing the model back from package system.
> 
> Indeed this call chain does not attach bytecode, but is it needed? At least none of our v2 e2e tests failed which should take a similar code path.
> 
> If we really need to attach the bytecode to the model for running extended checks, we should add an experiment flag for doing so at the end of the compilation pipeline, and leverage the already existing code for this which I pointed to in the comment below (i.e. pull this existing code into the compiler driver and use the experiment flag wherever we need to have the attached bytecode).

The function [build_all](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/third_party/move/tools/move-package/src/compilation/build_plan.rs#L210) is used for compilation, docgen and abigen, which currently uses the model created [here](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L750). 


We need to attach the compiled module because V2 model does not have the compiled module but the extended check [here](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/aptos-move/framework/src/extended_checks.rs#L567) needs it.

wrwg commented:
> > I'm not 100% sure what the purpose of the PR is. We already use the v2 model in extended checks. Its created here:
> > https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L1113-L1124
> > 
> > ... which is called here:
> > https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/build_plan.rs#L139
> > 
> > ... which in turn is called from aptos-move `BuiltPackage::build` which calls extended checks with the model as created by the v2 compiler:
> > https://github.com/aptos-labs/aptos-core/blob/219d461f7d4e0366bb802ac6282e6f0d8b7266ce/aptos-move/framework/src/built_package.rs#L207-L212
> > 
> > So this was already running in many e2e tests, and there is no building of the model twice since I fixed this a while ago and implemented passing the model back from package system.
> > Indeed this call chain does not attach bytecode, but is it needed? At least none of our v2 e2e tests failed which should take a similar code path.
> > If we really need to attach the bytecode to the model for running extended checks, we should add an experiment flag for doing so at the end of the compilation pipeline, and leverage the already existing code for this which I pointed to in the comment below (i.e. pull this existing code into the compiler driver and use the experiment flag wherever we need to have the attached bytecode).
> 
> The function [build_all](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/third_party/move/tools/move-package/src/compilation/build_plan.rs#L210) is used for compilation, docgen and abigen, which currently uses the model created [here](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L750).
> 
> We need to attach the compiled module because V2 model does not have the compiled module but the extended check [here](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/aptos-move/framework/src/extended_checks.rs#L567) needs it.

Ah, I see. Then you need to leverage the model as returned here, the `Option<GlobalEnv>` should be set in the v2 compiler path. You may also need to add a new experiment ATTACH_BYTECODE_TO_MODEL if this is needed/

https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L640

rahxephon89 commented:
> > > I'm not 100% sure what the purpose of the PR is. We already use the v2 model in extended checks. Its created here:
> > > https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L1113-L1124
> > > 
> > > ... which is called here:
> > > https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/build_plan.rs#L139
> > > 
> > > ... which in turn is called from aptos-move `BuiltPackage::build` which calls extended checks with the model as created by the v2 compiler:
> > > https://github.com/aptos-labs/aptos-core/blob/219d461f7d4e0366bb802ac6282e6f0d8b7266ce/aptos-move/framework/src/built_package.rs#L207-L212
> > > 
> > > So this was already running in many e2e tests, and there is no building of the model twice since I fixed this a while ago and implemented passing the model back from package system.
> > > Indeed this call chain does not attach bytecode, but is it needed? At least none of our v2 e2e tests failed which should take a similar code path.
> > > If we really need to attach the bytecode to the model for running extended checks, we should add an experiment flag for doing so at the end of the compilation pipeline, and leverage the already existing code for this which I pointed to in the comment below (i.e. pull this existing code into the compiler driver and use the experiment flag wherever we need to have the attached bytecode).
> > 
> > 
> > The function [build_all](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/third_party/move/tools/move-package/src/compilation/build_plan.rs#L210) is used for compilation, docgen and abigen, which currently uses the model created [here](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L750).
> > We need to attach the compiled module because V2 model does not have the compiled module but the extended check [here](https://github.com/aptos-labs/aptos-core/blob/dd40342cd9d9e5aa1c990f530e062d29b748104b/aptos-move/framework/src/extended_checks.rs#L567) needs it.
> 
> Ah, I see. Then you need to leverage the model as returned here, the `Option<GlobalEnv>` should be set in the v2 compiler path. You may also need to add a new experiment ATTACH_BYTECODE_TO_MODEL if this is needed/
> 
> https://github.com/aptos-labs/aptos-core/blob/d6483636c325f76c5207bb20f6160f0f395cba39/third_party/move/tools/move-package/src/compilation/compiled_package.rs#L640

Yes, I added the experiment flag in this [commit](https://github.com/aptos-labs/aptos-core/pull/12954/commits/107036c63bff5967e0f9d441705f7a5b03b50fb1)
